x/src/lint/engine/runner.rs

```diff
@@ -0,0 +1,183 @@
+
+            'outer: for file_ctx in file_ctxs {
+                let linters_to_run = content_linters
+                    .iter()
+                    .copied()
+                    .filter_map(|linter| match linter.pre_run(&file_ctx) {
+                        Ok(RunStatus::Executed) => Some(Ok(linter)),
+                        Ok(RunStatus::Skipped(reason)) => {
+                            let source = file_ctx.source(linter.name());
+                            skipped.push((source, reason));
+                            None
+                        }
+                        Err(err) => Some(Err(err)),
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                if linters_to_run.is_empty() {
+                    // No linters to run for this file -- no point loading it.
+                    continue;
+                }
+
+                // Load up the content for this file.
+                let content_ctx = file_ctx.load()?;
+
+                for linter in linters_to_run {
+                    let source = content_ctx.source(linter.name());
+                    let mut formatter = LintFormatter::new(source, &mut messages);
+
+                    match linter.run(&content_ctx, &mut formatter)? {
+                        RunStatus::Executed => {
+                            // Yay! Lint ran successfully.
+                        }
+                        RunStatus::Skipped(reason) => {
+                            skipped.push((source, reason));
+                        }
+                    }
+
+                    if self.config.fail_fast && !messages.is_empty() {
+                        // At least one issue was found.
+                        break 'outer;
+                    }
+                }
+            }
+        }
+
+        Ok(LintResults { skipped, messages })
+    }
+
+    // ---
+    // Helper methods
+    // ---
+
+    fn get_file_list(&self) -> Result<Vec<&Path>> {
+        // TODO: abstract out SCM and command-running functionality.
+        let stdout = self.files_stdout()?;
+
+        Ok(stdout
+            .split(|b| b == &0)
+            .filter_map(|s| {
+                // TODO: make global exclusions configurable.
+                if !s.is_empty() && !s.starts_with(b"testsuite/libra-fuzzer/artifacts/") {
+                    // Well, this only works on Unix for now. Would be cool to get it working on
+                    // Windows at some point too, but it needs to be done with some care. See
+                    // https://www.mercurial-scm.org/wiki/EncodingStrategy and
+                    // https://en.wikipedia.org/wiki/Mojibake.
+                    use std::os::unix::ffi::OsStrExt;
+                    let s = OsStr::from_bytes(s);
+                    Some(Path::new(s))
+                } else {
+                    None
+                }
+            })
+            .collect())
+    }
+
+    fn files_stdout(&self) -> Result<&[u8]> {
+        self.files_stdout
```

Comment:
Such a great use of once_cell, I wonder, does it make sense to actually have this cell capture the output of the file-listing rather than the raw output from git?

Reply:
Heh, I thought about that -- but that would involve creating lots of `PathBuf` instances and doing lots of allocations. In this case

Reply:
I've added a comment explaining the options.

Reply:
Yeah that's true, I suppose the question comes down to is: Is it more expensive to do the parsing into `Path`s each time (idk how expensive the conversion is) or to do lots of little allocs. 🤷‍♂ At this point though its 100% a micro optimization and probably not worth digging into it more yet. Lets leave it how you have it.

Reply:
I did some measurements and it looks like `cargo x lint` ends up being around 10% faster overall with this.

Reply:
The conversion to `Path` is free since on Unixes it's just a wrapper around a `[u8]` -- the thing that will have to be recomputed is searching for null bytes in the output, which I imagine is extremely fast.